### PR TITLE
Connect search to backend

### PR DESCRIPTION
**Pull request by kevingreer**
This PR doesn't have an indicator while searching, and it doesn't have an indicator that there were no results

### CODE REVIEW

File: Podcast/SearchAllEndpointRequest.swift [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+//  Created by Kevin Greer on 3/14/17.
+//  Copyright © 2017 Cornell App Development. All rights reserved.
+//
+
+import UIKit
+import SwiftyJSON
+
+class SearchAllEndpointRequest: EndpointRequest {
+    
+    init(query: String, offset: Int, max: Int) {
```
Comment (dennisfedorko) [inline Podcast/SearchAllEndpointRequest.swift]: add parameters as class properties when you make endpoint requests 

not super necessary, but its a useful convention we should maintain

File: Podcast/SearchAllEndpointRequest.swift [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+        httpMethod = .get
+        
+        queryParameters = ["offset": offset, "max": max]
+    }
+    
+    override func processResponseJSON(_ json: JSON) {
+        let episodes = json["data"]["episodes"].map{ episodeJson in Episode(json: episodeJson.1) }
+        let series = json["data"]["series"].map{ episodeJson in Series(json: episodeJson.1) }
+        let users = json["data"]["users"].map{ episodeJson in User(json: episodeJson.1) }
+        let results: [SearchType: [Any]] = [.episodes: episodes, .series: series, .people: users]
```
Comment (dennisfedorko) [inline Podcast/SearchAllEndpointRequest.swift]: processedResponseValue = [.episodes: episodes, .series: series, .people: users]

you don't need to specify the type

File: Podcast/SearchEpisodesEndpointRequest.swift [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+//  Created by Kevin Greer on 3/18/17.
+//  Copyright © 2017 Cornell App Development. All rights reserved.
+//
+
+import UIKit
+import SwiftyJSON
+
+class SearchEpisodesEndpointRequest: EndpointRequest {
+    
+    init(query: String, offset: Int, max: Int) {
```
Comment (dennisfedorko) [inline Podcast/SearchEpisodesEndpointRequest.swift]: make params properties

File: Podcast/SearchEpisodesEndpointRequest.swift [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+        
+        path = "/search/episodes/\(query.addingPercentEncoding(withAllowedCharacters: .urlHostAllowed) ?? query)"
+        
+        httpMethod = .get
+        
+        queryParameters = ["offset": offset, "max": max]
+    }
+    
+    override func processResponseJSON(_ json: JSON) {
+        processedResponseValue = json["data"]["episodes"].map{ episodeJson in Episode(json: episodeJson.1) }
```
Comment (dennisfedorko) [inline Podcast/SearchEpisodesEndpointRequest.swift]: episodeJSON

File: Podcast/SearchSeriesEndpointRequest.swift [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+//  Created by Kevin Greer on 3/18/17.
+//  Copyright © 2017 Cornell App Development. All rights reserved.
+//
+
+import UIKit
+import SwiftyJSON
+
+class SearchSeriesEndpointRequest: EndpointRequest {
+    
+    init(query: String, offset: Int, max: Int) {
```
Comment (dennisfedorko) [inline Podcast/SearchSeriesEndpointRequest.swift]: same issue as above

File: Podcast/SearchSeriesEndpointRequest.swift [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+        
+        path = "/search/series/\(query.addingPercentEncoding(withAllowedCharacters: .urlHostAllowed) ?? query)"
+        
+        httpMethod = .get
+        
+        queryParameters = ["offset": offset, "max": max]
+    }
+    
+    override func processResponseJSON(_ json: JSON) {
+        processedResponseValue = json["data"]["series"].map{ episodeJson in Series(json: episodeJson.1) }
```
Comment (dennisfedorko) [inline Podcast/SearchSeriesEndpointRequest.swift]: episodeJSON

File: Podcast/SearchTableViewController.swift [hunk 7]
```diff
@@ -13,10 +13,24 @@ enum SearchType {
     case series
     case people
     case tags
+    
+    var string: String {
```
Comment (dennisfedorko) [inline Podcast/SearchTableViewController.swift]: make this a func, toString() -> String

File: Podcast/SearchTableViewController.swift [hunk 8]
```diff
@@ -86,6 +104,11 @@ class SearchTableViewController: UITableViewController {
     override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
         cellDelegate?.searchTableViewController(controller: self, didTapSearchResultOfType: searchType, index: indexPath.row)
     }
+        
+    func fetchData(_ completionHandler: ((Void) -> Void)?) {
```
Comment (dennisfedorko) [inline Podcast/SearchTableViewController.swift]: fetchData(completion: (() -> ())?) {...}

File: Podcast/SearchUsersEndpointRequest.swift [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+//  Created by Kevin Greer on 3/18/17.
+//  Copyright © 2017 Cornell App Development. All rights reserved.
+//
+
+import UIKit
+import SwiftyJSON
+
+class SearchUsersEndpointRequest: EndpointRequest {
+    
+    init(query: String, offset: Int, max: Int) {
```
Comment (dennisfedorko) [inline Podcast/SearchUsersEndpointRequest.swift]: same issue as above

File: Podcast/TabbedPageViewController.swift [hunk 10]
```diff
@@ -36,6 +36,13 @@ class TabbedPageViewController: UIViewController, UIPageViewControllerDataSource
         .series: [],
         .people: [],
         .tags: []]
+    
+    let PageSize = 20
```
Comment (dennisfedorko) [inline Podcast/TabbedPageViewController.swift]: pageSize

File: Podcast/TabbedPageViewController.swift [hunk 11]
```diff
@@ -167,25 +146,95 @@ class TabbedPageViewController: UIViewController, UIPageViewControllerDataSource
     
     
     func updateSearchResults(for searchController: UISearchController) {
-        guard let searchText = searchController.searchBar.text else { return }
-        updateCurrentViewControllerTableView()
+        guard let searchText = searchController.searchBar.text, searchText != "" else { return }
+        print("updateSearchResults", searchText)
+        self.searchText = searchText
+        self.sectionOffsets = [.episodes: 0, .series: 0, .people: 0, .tags: 0]
```
Comment (dennisfedorko) [inline Podcast/TabbedPageViewController.swift]: why do you need self here?

File: Podcast/TabbedPageViewController.swift [hunk 12]
```diff
@@ -167,25 +146,95 @@ class TabbedPageViewController: UIViewController, UIPageViewControllerDataSource
...
+    
+    func fetchData(type: SearchType, query: String, offset: Int, max: Int, append: Bool) {
+        
+        System.endpointRequestQueue.cancelAllEndpointRequestsOfType(type: SearchEpisodesEndpointRequest.self)
+        System.endpointRequestQueue.cancelAllEndpointRequestsOfType(type: SearchUsersEndpointRequest.self)
+        System.endpointRequestQueue.cancelAllEndpointRequestsOfType(type: SearchSeriesEndpointRequest.self)
+        System.endpointRequestQueue.cancelAllEndpointRequestsOfType(type: SearchAllEndpointRequest.self)
+        
+        if append {
+            switch type {
```
Comment (dennisfedorko) [inline Podcast/TabbedPageViewController.swift]: why don't you have a .all type so you don't need to use append and can flatten this logic?

File: Podcast/SearchAllEndpointRequest.swift (hunk 2 above)
Comment (kevingreer) [reply Podcast/SearchAllEndpointRequest.swift]: You do so it knows .episodes refers to SearchType.episodes

File: Podcast/TabbedPageViewController.swift [hunk 13]
```diff
@@ -167,25 +146,91 @@ class TabbedPageViewController: UIViewController, UIPageViewControllerDataSource
     
     
     func updateSearchResults(for searchController: UISearchController) {
-        guard let searchText = searchController.searchBar.text else { return }
-        updateCurrentViewControllerTableView()
+        guard let searchText = searchController.searchBar.text, searchText != "" else { return }
+        print("updateSearchResults", searchText)
```
Comment (dennisfedorko) [inline Podcast/TabbedPageViewController.swift]: delete print statement

**Pull request merged**